app/WebParsing/UtsgJsonParser.hs

```diff
@@ -42,17 +44,54 @@ insertAllMeetings org = do
     manager <- liftIO $ newManager tlsManagerSettings
     response <- liftIO $ httpLbs request' manager
     let respBody = responseBody response
+    return respBody
 
-    -- decode the response
+-- Get the page number, page size and total number of courses from response
+getPageInfo :: ByteString -> Maybe (Int, Int, Int)
+getPageInfo respBody = do
+    jsonValue <- decode respBody
+    case jsonValue of 
+      Nothing -> return (-1, -1, -1)
+      Just json -> do 
+        flip parseMaybe json $ \obj -> do
+          payload :: Object <- obj .: "payload"
+          pageableCourse :: Object <- payload .: "pageableCourse"
+          page :: Int <- pageableCourse .: "page"
+          pageSize :: Int <- pageableCourse .: "pageSize"
+          totalCourses :: Int <- pageableCourse .: "total"
+          return (page, pageSize, totalCourses)
+          
+-- Helper function to insert courses for a page of a HTTP response
+insertCourses :: ByteString -> SqlPersistM ()
+insertCourses respBody = do
     let meetings :: Maybe DBList = decode respBody
     case meetings of
       Nothing -> return ()
-      Just dblist ->  mapM_ insertMeeting $ flattenDBList dblist
+      Just dblist -> do mapM_ insertMeeting $ flattenDBList dblist
 
 -- | Helper function to flatten the list of DB Objects
 flattenDBList :: DBList -> [MeetTime]
 flattenDBList (DBList meetings) = concatMap (\(DB meetTimes) -> meetTimes) meetings
 
+-- | insert/update all the data into the Meeting and Times schema by creating and sending
+--   the http request to Artsci Timetable and then parsing the JSON response
+insertAllMeetings :: Int -> Int -> Int -> SqlPersistM ()
+insertAllMeetings page pageSize totalCourses = do
+    if (page - 1) * pageSize > totalCourses  -- base case
+      then liftIO $ putStrLn "All courses have been parsed."
+      else do
+        liftIO $ print $ "Parsing results for page " ++ show page ++ " of " ++ show totalPages
+        respBody <- liftIO $ makeRequest page
+        let pageInfo :: Maybe (Int, Int, Int) = getPageInfo respBody
+        case pageInfo of 
+          Nothing -> return ()
+          Just (page', pageSize', totalCourses') -> do
+            insertCourses respBody 
+            insertAllMeetings (page' + 1) pageSize' totalCourses'
+        where
+          totalPages :: Integer
+          totalPages = fromIntegral (ceiling (fromIntegral totalCourses / fromIntegral pageSize :: Double) :: Int)
```

Comment:
keep the type annotation as `Int` here; you can remove the `fromIntegral` and `:: Int`, and let Haskell's type inference do the rest for you.